xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java

```diff
@@ -0,0 +1,281 @@
+        int curScore = getFilterChainMatchScore(filterChain.getFilterChainMatch(), localInetAddr);
+        if (curScore > 0 && curScore > highestScore) {
+          bestMatch = filterChain;
+          highestScore = curScore;
+        }
+      }
+      if (bestMatch != null) {
+        return bestMatch.getDownstreamTlsContext();
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Computes a score for a match of filterChainMatch with channel.
+   *
+   * <p>-1 => mismatch (of port or IP address etc)
+   *
+   * <p>0 => unimplemented (types or logic etc)
+   *
+   * <p>1 => filterChainMatch is null, so nothing to match.
+   *
+   * <p>value > 2 indicates some kind of IP address match.
+   */
+  private static int getFilterChainMatchScore(
+      FilterChainMatch filterChainMatch, InetSocketAddress localInetAddr) {
+    if (filterChainMatch == null) {
+      return 1;
+    }
+    if (filterChainMatch.getDestinationPort() != localInetAddr.getPort()) {
+      return -1;
+    }
+    return getIpAddressAndRangesMatchScore(
+        localInetAddr.getAddress(), filterChainMatch.getPrefixRanges());
+  }
+
+  /**
+   * Computes a score for IP address match against a list of CidrRange called only after ports have
+   * matched, so a minimum score of 2 is returned to indicate a match or else -1 for a mismatch.
+   *
+   * <p>-1 => mismatch.
+   *
+   * <p>2 => prefixRanges is null/empty, so no IP address to match.
+   *
+   * <p>value > 2 indicates some kind of IP address match.
+   */
+  private static int getIpAddressAndRangesMatchScore(
+      InetAddress localAddress, List<CidrRange> prefixRanges) {
+    if (prefixRanges == null || prefixRanges.isEmpty()) {
+      return 2;
+    }
+    int highestScore = -1;
+    for (CidrRange cidrRange : prefixRanges) {
+      int curScore = getIpAddressMatchScore(localAddress, cidrRange);
+      if (curScore > highestScore) {
+        highestScore = curScore;
+      }
+    }
+    return highestScore;
+  }
+
+  /**
+   * Computes a score for IP address to CidrRange match.
+   *
+   * <p>-1 => mismatch
+   *
+   * <p>0 => unimplemented (prefixLen < 32 logic)
+   *
+   * <p>4 => match because cidrRange is IPANY_ADDRESS (0.0.0.0)
+   *
+   * <p>8 => exact match
+   */
+  private static int getIpAddressMatchScore(InetAddress localAddress, CidrRange cidrRange) {
+    if (cidrRange.getPrefixLen() == 32) {
+      try {
+        InetAddress cidrAddr = InetAddress.getByName(cidrRange.getAddressPrefix());
+        if (cidrAddr.isAnyLocalAddress()) {
+          return 4;
+        }
+        if (cidrAddr.equals(localAddress)) {
+          return 8;
+        }
+        return -1;
+      } catch (UnknownHostException e) {
+        logger.log(Level.SEVERE, "cidrRange address parsing", e);
+      }
+    }
+    // TODO(sanjaypujare): implement CIDR logic to match prefixes if needed
+    return 0;
+  }
+
+  /** Shutdown this instance and release resources. */
+  public void shutdown() {
+    logger.log(Level.INFO, "Shutdown");
```

Comment:
nit: this will be printed as just "Shutdown", without anything else. It doesn't give any useful information.

Also, I doubt there should be any log messages above INFO level as anything above (or equal) INFO level will be printed out by default.

Reply:
So all the *dsLoadBalancer.shutdown() (e.g. `CdsLoadBalancer.shutdown`) have this:
```
  public void shutdown() {
    logger.log(XdsLogLevel.INFO, "Shutdown");
...
}
```
Agreed `XdsLogLevel.INFO` is not the same as `Level.INFO`  but they all say "Shutdown" so being consistent. I am changing from `INFO` to `FINER` .

Reply:
xDS policies (e.g., CdsLoadBalancer) uses `XdsLogger`, which prepends a unique `LogId` to the log message. So logging "Shutdown" really would print out "[cds-lb<ID>] Shutdown". But here, when seeing a "Shutdown" log message, we have no clue (unless the class name is shown) what is being shut down. Not a big issue.

Reply:
each log line output has the classname and method as the prefix so users know where this message is coming from.